feat: add no show setting in cal video

dhairyashiil opened the pull request:
## What does this PR do?

- Fixes #24319 
- Fixes CAL-6532 

## Visual Demo (For contributors especially)


https://github.com/user-attachments/assets/df2962ec-bd57-45db-bd78-eb74bdf3de5b

dhairyashiil committed:
feat: add no show setting in cal video

pallava-joshi reviewed (changes requested):
please address code-rabbit suggestion

dhairyashiil committed:
Use normalized calVideoSettings in booking flow and add separate backfill migration for Cal Video no-show flags

dhairyashiil committed:
resolve merge conflicts

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into feat/cal-video-no-show-setting-2

pallava-joshi reviewed (dismissed):
got some cubic suggestions

dhairyashiil committed:
resolve merge conflicts

Udit-takkar commented on packages/features/eventtypes/components/locations/CalVideoSettings.tsx:
```diff
@@ -16,12 +18,32 @@ import { Tooltip } from "@calcom/ui/components/tooltip";
+    { eventTypeId: eventTypeId ?? undefined },
+    { enabled: !!eventTypeId }
+  );
+
```
There could also be team wide webhooks, org wide webhooks

Udit-takkar reviewed (commented):
I think to reduce complexity here https://github.com/calcom/cal.com/pull/24351#discussion_r2479377287 we skip the check for webhook and simply mention somewhere that if you have "after hosts didn't join cal video", "after guests didn't join cal video" webhook enabled then we would automatically track their no show. let's discuss first

dhairyashiil committed:
address review comments

dhairyashiil committed:
resolved merge conflicts

dhairyashiil committed:
better code quality

dhairyashiil committed:
Use findFirst Instead of findMany

dhairyashiil commented:
### Video: 

https://github.com/user-attachments/assets/32618dfc-d361-41e1-9e49-6a8826abbe11

dhairyashiil committed:
fix failing type check

dhairyashiil commented on packages/prisma/migrations/20251014120000_backfill_cal_video_no_show_flags/migration.sql:
```diff
@@ -0,0 +1,35 @@
+-- Backfill CalVideoSettings no-show flags based on existing automations
+
```
removed backfill migration file

dhairyashiil committed:
Merge branch 'main' into feat/cal-video-no-show-setting-2

dhairyashiil committed:
addressed cubics comments

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into feat/cal-video-no-show-setting-2

pallava-joshi committed:
Merge branch 'main' into feat/cal-video-no-show-setting-2

dhairyashiil committed:
fix: use valid URL for internal no-show webhooks to pass schema validation

Previously, automatic no-show tracking used subscriberUrl: '' which violated
the z.string().url() constraint in the task schema, causing tasks to abort.

Now uses a special internal URL (https://internal.cal.com/no-webhook) that:
- Passes Zod schema validation
- Is detected by task handlers to skip actual webhook sending
- Still marks no-shows in database as intended

This fixes the automatic no-show tracking feature when no webhooks are configured.

dhairyashiil committed:
Merge branch 'feat/cal-video-no-show-setting-2' of https://github.com/dhairyashiil/cal.com into feat/cal-video-no-show-setting-2

Udit-takkar committed:
Merge branch 'main' into feat/cal-video-no-show-setting-2

Udit-takkar reviewed (commented):
https://github.com/user-attachments/assets/580160ea-a876-4145-b76a-0600d7ebd129

There is some bug when you select cal video and click advanced setting then event type update api is automatically triggered. Not sure if it related to the PR

Udit-takkar reviewed (commented):
https://github.com/user-attachments/assets/45cd46fc-3291-4210-900c-057d3ef909fc

Cal video settings is not saved

Udit-takkar commented:
We should also add some unit tests

dhairyashiil committed:
fix: Cal Video no-show settings not persisting after save

- Add enableAutomaticNoShowTrackingForHosts/Guests to form validation schema
- Add fields to eventTypeRepository select queries for proper data fetching
- Add fields to tRPC input validation schema
- Add fields to CalVideoSettings repository type definitions
- Add fields to create handler for new event types

The issue was caused by missing fields in multiple validation layers,
particularly in the form-level Zod schema which was stripping the fields
before submission. Now the settings properly save to DB and persist after reload.

dhairyashiil committed:
fix: add authorization checks to hasNoShowForEventType endpoint

- Add comprehensive authorization checks before accessing webhook data
- Verify user has access to event type (direct ownership, assigned, or team member)
- Optimize query by checking direct ownership first (fastest path)
- Prevent unauthorized users from probing other teams' webhook configurations

Addresses security review comment about missing authorization validation.

dhairyashiil committed:
refactor: use explicit flag for automatic no-show tracking

Replace fake webhook URLs with isAutomaticTrackingOnly flag for clarity.
When Cal Video automatic tracking is enabled without webhooks, we now:
- Set isAutomaticTrackingOnly: true flag
- Use empty subscriberUrl instead of fake URL
- Check the flag instead of URL patterns

This makes the code more maintainable and addresses review feedback
about confusing dummy webhook objects.

dhairyashiil committed:
test: add unit tests for Cal Video no-show tracking toggles

Add comprehensive tests for automatic no-show tracking feature:
- Test enableAutomaticNoShowTrackingForHosts toggle
- Test enableAutomaticNoShowTrackingForGuests toggle
- Test both toggles enabled/disabled scenarios
- Test webhook precedence over automatic tracking
- Test location validation (Cal Video only)
- Test edge cases (null settings, isDryRun, empty location)

10 essential tests with full coverage of the toggle feature.

dhairyashiil committed:
Merge branch 'feat/cal-video-no-show-setting-2' of https://github.com/dhairyashiil/cal.com into feat/cal-video-no-show-setting-2

dhairyashiil committed:
fix: add isAutomaticTrackingOnly to Zod schema and optimize calVideoSettings select

Address reviewer comments:

1. Add isAutomaticTrackingOnly to ZSendNoShowWebhookPayloadSchema
   - Update schema.ts to include the optional boolean field
   - Update common.ts prepareNoShowTrigger to extract and return the field
   - Update triggerHostNoShow.ts and triggerGuestNoShow.ts to use properly typed field
   - Remove unsafe 'as any' type assertions

2. Optimize calVideoSettings select in confirm.handler.ts
   - Only select required fields (enableAutomaticNoShowTrackingForHosts, enableAutomaticNoShowTrackingForGuests)
   - Remove unnecessary spreading and redirectUrlOnExit handling
   - Follows project guideline to reduce performance overhead

Fixes type check errors and improves code quality.

dhairyashiil commented:
1. Authorization check - Fixed
Added authorization checks. Includes user, team, and org checks.

2. Cal video settings not saved - Fixed
The issue was that the new fields were missing from multiple Zod schemas, causing them to be stripped during validation. Added `enableAutomaticNoShowTrackingForHosts` and `enableAutomaticNoShowTrackingForGuests` to all required schemas.

3. Unit tests - Added
Created `scheduleNoShowTriggers.test.ts`.

4. "Internal webhook" / dummy webhooks confusion - Refactored
Changed to use an explicit `isAutomaticTrackingOnly` flag instead:
Added `isAutomaticTrackingOnly: z.boolean().optional()` to the Zod schema
Tasks check this flag and skip webhook sending while still marking no-shows in DB

@Udit-takkar please check 🙏🏼

Udit-takkar reviewed (dismissed):
LGTM

Udit-takkar committed:
Merge branch 'main' into feat/cal-video-no-show-setting-2

devin-ai-integration[bot] committed:
Merge upstream/main into feat/cal-video-no-show-setting-2

Resolved merge conflicts in:
- apps/web/modules/event-types/components/locations/CalVideoSettings.tsx
- apps/web/modules/event-types/components/locations/Locations.tsx
- packages/features/bookings/lib/service/RegularBookingService.ts
- packages/features/eventtypes/lib/types.ts
- packages/features/tasker/tasks/triggerNoShow/triggerGuestNoShow.ts
- packages/features/tasker/tasks/triggerNoShow/triggerHostNoShow.ts

Also updated:
- packages/features/eventtypes/lib/schemas.ts (added no-show tracking fields)
- packages/features/bookings/lib/handleNewBooking/scheduleNoShowTriggers.ts (added version field)

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
Merge upstream/main into feat/cal-video-no-show-setting-2

Co-Authored-By: unknown <>

eunjae-lee reviewed (changes requested):
CI fails
